### PR TITLE
Initial implementation

### PR DESCRIPTION
**Pull request by bajtos**
Features implemented:
- SET key value (ttl)
- GET key value
- ITERATE KEYS

Limitations & features not available:
- EXPIRE
- QUERY TTL
- ITERATE KEYS returns only first 500 records

Connect to strongloop-internal/scrum-loopback#1126
See also the spike - https://github.com/strongloop/loopback-datasource-juggler/pull/1129

@superkhau please review

### CODE REVIEW

File: lib/kv-extreme-scale.js [hunk 1]
```diff
@@ -0,0 +1,370 @@
...
+var BinaryPacker = connectorCore.BinaryPacker;
+var ModelKeyComposer = connectorCore.ModelKeyComposer;
+
+exports.initialize = function initializeDataSource(dataSource, callback) {
+  var settings = dataSource.settings;
+
+  dataSource.connector = new ExtremeScaleKVConnector(settings, dataSource);
+
+  if (!callback) return;
+  return process.nextTick(callback);
```
Comment (superkhau) [inline lib/kv-extreme-scale.js]: do we need the return in front of proces.nextTick?

File: lib/kv-extreme-scale.js [hunk 2]
```diff
@@ -0,0 +1,370 @@
...
+ *
+ * @header ExtremeScaleKeyValueConnector(settings, dataSource)
+ */
+function ExtremeScaleKVConnector(settings, dataSource) {
+  Connector.call(this, 'kv-extreme-scale', settings);
+  this.dataSource = dataSource;
+
+  debug('Connector settings', settings);
+
+  // set key expiration based on LastUpdateTime with Optimistic locking
```
Comment (superkhau) [inline lib/kv-extreme-scale.js:41]: nit: lowercase O for optimistic

File: lib/kv-extreme-scale.js [hunk 3]
```diff
@@ -0,0 +1,370 @@
...
+  this._url = settings.url + '.LUT.O';
+
+  this._queryUrl = this._url
+    .replace(/\/v1\/grids\//, '/v1/query/')
+    .replace(/\/$/, '');
```
Comment (superkhau) [inline lib/kv-extreme-scale.js]: Maybe add some comments to explain these regexes?

File: lib/kv-extreme-scale.js [hunk 4]
```diff
@@ -0,0 +1,370 @@
...
+
+  this._strictSSL = settings.strictSSL;
+  this._cookieJar = request.jar();
+  this._packer = new BinaryPacker();
+  this.DataAccessObject = dataSource.juggler.KeyValueAccessObject;
+};
+
```
Comment (superkhau) [inline lib/kv-extreme-scale.js:55]: remove empty line to group with constructor?

File: lib/kv-extreme-scale.js [hunk 5]
```diff
@@ -0,0 +1,370 @@
...
+ExtremeScaleKVConnector.prototype._clear = function(composedKey, callback) {
+  var requestOptions = {
+    uri: '/' + encodeURIComponent(composedKey),
+    method: 'DELETE',
+  };
+
+  this.request(requestOptions, callback);
+};
+
+/*
```
Comment (superkhau) [inline lib/kv-extreme-scale.js]: /**

File: lib/kv-extreme-scale.js [hunk 6]
```diff
@@ -0,0 +1,370 @@
...
+      }
+
+      callback(null, parsed.key);
+    });
+  }
+
+  function reportEnd(callback) {
+    setImmediate(function() {
+      callback();
+    });
```
Comment (superkhau) [inline lib/kv-extreme-scale.js]: Does `setImmediate(callback);` work?

File: lib/kv-extreme-scale.js [hunk 7]
```diff
@@ -0,0 +1,370 @@
...
+
+    if (res.statusCode / 100 !== 2) {
+      debug('err code=%s body=%j',
+        res.statusCode,
+        (body instanceof Buffer ? body.toString() : body));
+      return callback(createErrorFromResponse(res, body));
+    }
+
+    debug('success (%s), body=%s', res.statusCode, body);
+    callback(err, res, body);
```
Comment (superkhau) [inline lib/kv-extreme-scale.js]: `null, res, body` since you already checked for err on 319?

File: lib/kv-extreme-scale.js [hunk 8]
```diff
@@ -0,0 +1,370 @@
...
+
+function createErrorFromResponse(res, body) {
+  debug('%s %s %j', res.statusCode, res.statusMessage, body);
+  if (body instanceof Buffer) {
+    body = body.toString();
+  }
+
+  // We cannot rely on Content-Type response header as the server
+  // returns "application/atom+xml" (??)
+  var looksLikeJsonStrig = typeof body === 'string' &&
```
Comment (superkhau) [inline lib/kv-extreme-scale.js]: Strig -> String

File: lib/kv-extreme-scale.js [hunk 9]
```diff
@@ -0,0 +1,370 @@
...
+    body[0] === '{' &&
+    body[body.length - 1] === '}';
+  if (looksLikeJsonStrig) {
```
Comment (superkhau) [inline lib/kv-extreme-scale.js]: ditto

File: package.json [hunk 10]
```diff
@@ -4,7 +4,9 @@
   "description": "The official WebSphere eXtreme Scale KeyValue connector for LoopBack",
   "main": "lib/kv-extreme-scale.js",
   "scripts": {
-    "posttest": "eslint . --ignore-path .gitignore"
+    "lint": "eslint . --ignore-path .gitignore",
```
Comment (superkhau) [inline package.json]: Is ignore-path required?

**Review by superkhau — DISMISSED** [review]

**Comment by superkhau** [comment]
Some minor fixes, LGTM. No more review needed.

File: lib/kv-extreme-scale.js (hunk 2 above)
Comment (bajtos) [reply lib/kv-extreme-scale.js:41]: Upper-case `O` is referring to `.O`  in the `.LUT.O` postfix.

File: lib/kv-extreme-scale.js (hunk 4 above)
Comment (bajtos) [reply lib/kv-extreme-scale.js:55]: I prefer to add an empty line after each method. In my view, the `inherits` line is a standalone statement that should stand out.

**Comment by bajtos** [comment]
@slnode test please

**Pull request merged**